components/layout_2020/quotes.rs

```diff
@@ -0,0 +1,406 @@
+        ("no", QuotesData(0xab, 0xbb, 0x2018, 0x2019)),
+        ("rw", QuotesData(0xab, 0xbb, 0x2018, 0x2019)),
+        ("fi", QuotesData(0x201d, 0x201d, 0x2019, 0x2019)),
+        ("he", QuotesData(0x201d, 0x201d, 0x2019, 0x2019)),
+        ("lag", QuotesData(0x201d, 0x201d, 0x2019, 0x2019)),
+        ("rn", QuotesData(0x201d, 0x201d, 0x2019, 0x2019)),
+        ("sn", QuotesData(0x201d, 0x201d, 0x2019, 0x2019)),
+        ("sv", QuotesData(0x201d, 0x201d, 0x2019, 0x2019)),
+        ("fr-CA", QuotesData(0xab, 0xbb, 0x201d, 0x201c)),
+        ("fr", QuotesData(0xab, 0xbb, 0xab, 0xbb)),
+        ("hy", QuotesData(0xab, 0xbb, 0xab, 0xbb)),
+        ("yav", QuotesData(0xab, 0xbb, 0xab, 0xbb)),
+        ("hu", QuotesData(0x201e, 0x201d, 0xbb, 0xab)),
+        ("ia", QuotesData(0x2018, 0x2019, 0x201c, 0x201d)),
+        ("nso", QuotesData(0x2018, 0x2019, 0x201c, 0x201d)),
+        ("ti-ER", QuotesData(0x2018, 0x2019, 0x201c, 0x201d)),
+        ("tn", QuotesData(0x2018, 0x2019, 0x201c, 0x201d)),
+        ("ja", QuotesData(0x300c, 0x300d, 0x300e, 0x300f)),
+        ("yue", QuotesData(0x300c, 0x300d, 0x300e, 0x300f)),
+        ("zh-Hant", QuotesData(0x300c, 0x300d, 0x300e, 0x300f)),
+        ("ka", QuotesData(0x201e, 0x201c, 0xab, 0xbb)),
+        ("nl", QuotesData(0x2018, 0x2019, 0x2018, 0x2019)),
+        ("nmg", QuotesData(0x201e, 0x201d, 0xab, 0xbb)),
+        ("pl", QuotesData(0x201e, 0x201d, 0xab, 0xbb)),
+        ("ro", QuotesData(0x201e, 0x201d, 0xab, 0xbb)),
+        ("shi", QuotesData(0xab, 0xbb, 0x201e, 0x201d)),
+        ("zgh", QuotesData(0xab, 0xbb, 0x201e, 0x201d)),
+        ("sr", QuotesData(0x201e, 0x201d, 0x2019, 0x2019)),
+        ("st", QuotesData(0x201c, 0x2019, 0x201c, 0x201d)),
+        ("tk", QuotesData(0x201c, 0x201d, 0x201c, 0x201d)),
+        ("uz", QuotesData(0x201c, 0x201d, 0x2019, 0x2018)),
+    ];
+    HashMap::from(input)
+}
+
+fn quotes_data_for_lang(lang: &str) -> QuotesData {
+    // All valid language codes are at least two bytes long.
+    if lang.len() < 2 {
+        return DEFAULT_QUOTES;
+    }
+
+    let quotes_map = get_quotes_map();
+
+    // Found an exact match for the requested lang.
+    if let Some(quotes_data) = quotes_map.get(lang) {
+        return *quotes_data;
+    }
+
+    // Try parsing lang as a Locale and canonicalizing the subtags, then see if
+    // we can match it with region or script subtags, if present, or just the
+    // primary language tag.
+    let maybe_loc = lang.parse::<Locale>();
+    if maybe_loc.is_err() {
+        return DEFAULT_QUOTES;
+    }
+
+    let loc = maybe_loc.unwrap();
+    let lang = loc.id.language.to_string();
+
+    if let Some(quotes_data) = quotes_map.get(lang.as_str()) {
+        return *quotes_data;
+    }
+
+    if let Some(region) = loc.id.region {
+        let lang_and_region = lang.clone() + "-" + region.as_str();
+        if let Some(quotes_data) = quotes_map.get(lang_and_region.as_str()) {
+            return *quotes_data;
+        }
+    }
```

Comment:
I think this can be done with map:

```suggestion
    if let Some(data) = loc.id.region.and_then(|region| quotes_data.get(format!("{lang}-{region")) {
        return data;
    }
```

Reply:
Done